storage.bs

```diff
@@ -94,37 +100,251 @@ function retrieveNextChunk(nextChunkInfo) {
 </dl>
 
-<p>This standard primarily concerns itself with <dfn export id=site-storage>storage</dfn>.
+<p>This standard primarily concerns itself with storage.
+
+
+<h3 id=storage-units>Storage units</h3>
+
+<p>A <dfn>storage type</dfn> is "<code>storage</code>" or "<code>session-storage</code>".
+
+<p>A <dfn>storage key</dfn> is an <a for=/>origin</a>. [[HTML]]
+
+<p class=XXX>This is expected to change, see
+<a href="https://privacycg.github.io/storage-partitioning/">Client-Side Storage Partitioning</a>.
+
+<p>A <dfn id=site-storage-unit>storage unit</dfn> is the least granular unit of storage, other than
+the user agent. It holds a <dfn for="storage unit">map</dfn>, which is a <a for=/>map</a> of
+<a for=/>ASCII strings</a> to <a>abstract storage buckets</a>.
+
+<p>A <dfn>storage map</dfn> is a <a for=/>map</a> of <a>storage keys</a> to <a>storage units</a>. It
+is initially empty.
+
+<p>To <dfn>obtain a storage unit</dfn>, given a <a for=/>storage map</a> <var>map</var>, an
+<a>environment settings object</a> <var>environment</var>, and a <a>storage type</a>
+<var>type</var>, run these steps:
+
+<ol>
+ <li><p>Let <var>key</var> be <var>environment</var>'s
+ <a for="environment settings object">origin</a>.
+
+ <li><p>If <var>key</var> is an <a>opaque origin</a>, then return failure.
+
+ <li><p>If the user has disabled storage, then return failure.
+
+ <li>
+  <p>If <var>map</var>[<var>key</var>] does not <a for=map>exist</a>, then:
+
+  <ol>
+   <li><p>Let <var>unit</var> be a new <a>storage unit</a>.
+
+   <li>
+    <p>Set <var>unit</var>'s <a for="storage unit">map</a>["<code>default</code>"] to the result of
+    <a>create a storage bucket</a> with <var>type</var>.
+
+    <p class="note">For now "<code>default</code>" is all that exists. See
+    <a href="https://github.com/whatwg/storage/issues/2">issue #2</a>.
+
+   <li><p>Set <var>map</var>[<var>key</var>] to <var>unit</var>.
+  </ol>
+
+ <li><p>Return <var>map</var>[<var>key</var>].
+</ol>
+
+<hr>
+
+<p>A <a for=/>user agent</a> holds a <dfn for="user agent">storage map</dfn>, which is a
+<a for=/>storage map</a>.
+
+<p>A browsing session holds a <dfn for="browsing session">storage map</dfn>, which is a
+<a for=/>storage map</a>.
+
+<p class=XXX>Browsing session is yet to be formally defined. For all intents and purposes it is a
+<a>top-level browsing context</a>, except that it survives the <a>top-level browsing context</a>
+being replaced due to a cross-origin opener policy.
+
+
+<h3 id=storage-endpoints>Storage endpoints</h3>
+
+<p>A <dfn export>storage endpoint</dfn> is a storage or session storage API that uses the
+infrastructure defined by this standard to keep track of its storage needs.
+
+<p>A <a>storage endpoint</a> has an <dfn for="storage endpoint">identifier</dfn>, which is a
+<a>storage identifier</a>.
+
+<p>A <a>storage endpoint</a> also has <dfn for="storage endpoint">types</dfn>, which is a
+<a for=/>set</a> of <a>storage types</a>.
```

Comment:
It's been requested to be able to use APIs with different backends. E.g., session IDB. This would allow for that.

Reply:
It'd be good to state that.

Reply:
Or just remove this extra complexity until such features arrive?

Reply:
That's an alternative. I wanted to have something that could address the full set of use cases everyone had in mind. As it doesn't really affect what's observable I think it's fine to leave it in for a bit until those various ideas get more concrete.

It would also be somewhat annoying to have to refactor various specifications when those features arrive, but maybe it's not too bad.